File: trees/binarySearch.go
package main

//O(log(n)) time and O(1) space
func BinarySearchIter(array []int, target int) int {
	l := 0
	r := len(array)-1
	for l<=r {
		m := (l+r)/2
		if array[m]==target {
			return m
		} else if array[m] > target {
			r = m-1
		}else{
			l = m+1
		}
	}
	return -1
}

func helper(array []int, target int, left int, right int) int {
	if left>right {
		return -1
	}
	mid := left+(right-left)/2
	if array[mid] == target {
		return mid
	} else if array[mid] > target {
		return helper(array, target, left, mid-1)
	} else {
		return helper(array, target, mid+1, right)
	}
}

//O(log(n)) time and O(log(n)) space
func BinarySearch(array []int, target int) int {
	return helper(array, target, 0, len(array))
}

File: trees/binarySearch.py
#O(log(n)) time and space O(1)
def binarySearchIter(array, target):
	l=0
	r=len(array)-1
	while l<=r:
		mid=l+(r-l)//2
		if array[mid]==target:
			return mid
		elif array[mid]>target:
			r=mid-1
		else:
			l=mid+1
	return -1

#O(log(n)) time and O(log(n)) space
def helper(array, target, left, right):
	if left>right:
		return -1
	mid=left+(right-left)//2
	if array[mid]==target:
		return mid
	elif array[mid]>target:
		return helper(array, target, left, mid-1)
	else:
		return helper(array, target, mid+1, right)

def binarySearch(array, target):
	return helper(array, target, 0, len(array)-1)
